client.go

```diff
@@ -135,27 +158,62 @@ type RequestParams struct {
 
 // Generic helper to execute HTTP requests using the net/http package
 //
-// nolint:unused, funlen, gocognit
-func (c *httpClient) doRequest(ctx context.Context, method, url string, params RequestParams) error {
+// nolint:funlen, gocognit, nestif
+func (c *Client) doRequest(ctx context.Context, method, endpoint string, params RequestParams, paginationMutator *func(*http.Request) error) error {
 	var (
 		req        *http.Request
 		bodyBuffer *bytes.Buffer
 		resp       *http.Response
 		err        error
 	)
 
-	for range httpDefaultRetryCount {
-		req, bodyBuffer, err = c.createRequest(ctx, method, url, params)
+	// Capture the original body into a buffer
+	originalBodyBuffer := new(bytes.Buffer)
+	if params.Body != nil {
+		reader, ok := params.Body.(io.Reader)
+		if !ok {
+			if c.debug && c.logger != nil {
+				c.logger.Errorf("failed to read body: params.Body is not an io.Reader")
+			}
+			return fmt.Errorf("failed to read body: params.Body is not an io.Reader")
```

Comment:
optional: We might be able to simplify encapsulate the log and error logic in a helper function.

e.g.
```go
func (c *Client) ErrorAndLogf(format string, args ...string) error { ... }
```